salt/utils/master.py

```diff
@@ -189,16 +189,16 @@ def _get_minion_grains(self, *minion_ids, **kwargs):
             cret = dict([(minion_id, mcache) for (minion_id, mcache) in six.iteritems(cached_grains) if mcache])
             missed_minions = [minion_id for minion_id in minion_ids if minion_id not in cret]
             log.debug('Missed cached minion grains for: %s', missed_minions)
-            if self.grains_fallback:
+            if self.grains_fallback and len(missed_minions) > 0:
```

Comment:
just do `if self.grains_fallback and missed_minions`

you do not need to check the length, if it is an empty list, it will evaluate as false, and if there are times in it it will evaluate as true

Reply:
this needs to be applied to the other changes as well